File: problems/C++/03. Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> dict(256, -1);
        int res = 0;
        int start = 0;

        for (int i = 0; i < s.length(); i++) {
            if (dict[s[i]] > -1) {
                start = max(start, dict[s[i]] + 1);
            }
            dict[s[i]] = i;
            res = max(res, i - start + 1);
        }
        return res;
    }
};

File: problems/Rust/03. Longest Substring Without Repeating Characters.rs
use std::collections::HashMap;
use std::cmp;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut map: HashMap<char, usize> = HashMap::new();
        let mut res = 0;
        let mut l: usize = 0;

        for (r, c) in s.char_indices() {
            if let Some(prev) = map.get(&c) {
                if *prev >= l {
                    l = *prev + 1;
                }
            }
            map.insert(c, r);
            res = cmp::max(res, r - l + 1);
        }
        res as i32
    }
}
